src/algorithms/diamondsquare.jl

```diff
@@ -34,10 +37,31 @@ DiamondSquare() = DiamondSquare(0.5)
 """
     _landscape!(mat, alg::DiamondSquare; kw...)
 
-    pass it on to the actual algorithm
+    Check if `mat` is the right size and already initialized.
+    If mat is not the correct size (DiamondSquare can only run on a lattice of size NxN where N = (2^n)+1 for integer n),
+    allocates the smallest lattice large enough to contain `mat` that can run DiamondSquare.
+    Will initialize `mat` to all zeros before running DiamondSquare if it is not initialized already.
 """
 function _landscape!(mat, alg::DiamondSquare; kw...) where {IT <: Integer}
-    diamondsquare!(mat, alg)
+
+    initialized::Bool = !any(isnan, mat)
+    rightSize::Bool = isPowerOfTwo(size(mat)[1]-1) && isPowerOfTwo(size(mat)[2]-1)
+    latticeSize::Int = size(mat)[1]
+
+    dsMat = mat
+    if !rightSize
+        dim1, dim2 = size(mat)
+        smallestContainingLattice::Int = 2^ceil(log2(max(dim1, dim2))) + 1
+        @warn "DiamondSquare() cannot be run on the input dimensions ($dim1 x $dim2),
+                and will instead run on the next smallest valid size ($smallestContainingLattice x $smallestContainingLattice).
+                This can slow performance as it involves additional memory allocation."
+        dsMat = zeros(smallestContainingLattice, smallestContainingLattice)
+    elseif !initialized
+        dsMat = zeros(latticeSize,latticeSize)
```

Comment:
This feels wrong to me. If the algorithm requires all values to be 0 at the beginning then all values should be set to 0 regardless of whether any values are `NaN`, right? Could you say where does the algorithm depend on the starting values?

Also, allocating an all new `dsMat` matrix here seems to defeat the purpose of having an in-place `landscape!` function in the first place.

Reply:
ah, in an all timer good coder moment i've noticed that  `diamond!` and `square!` were called in the wrong order (even though the order in which to call them is the name of the algorithm), which lead to interpolating from locations in the matrix that hadn't been assigned yet. seems to be fixed :facepalm:

Reply:
Great!"